### PR TITLE
Feature: File transfer in secure conversations 

### PR DESCRIPTION
**Pull request by vdveer**
* Add files to secure conversations by selecting in app (after enabling the preference)
* Share files (up to mms-sizelimit) and attach them to a secure chat

### CODE REVIEW

**Comment by vdveer** [comment]
@BastienLQ : I have moved things here, since the attachment refactor on the upstream made me create a new branch to work with.

**Comment by vdveer** [comment]
LABEL : donotmerge (i need to change the contenttype to a vendor-specific type)

File: src/org/smssecure/smssecure/ConversationActivity.java [hunk 1]
```diff
@@ -421,7 +427,7 @@ public void onMuted(final long until) {
           @Override
           protected Void doInBackground(Void... params) {
             DatabaseFactory.getRecipientPreferenceDatabase(ConversationActivity.this)
-                           .setMuted(recipients, until);
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/ConversationActivity.java]: Please avoid touching code unrelated to PR

File: src/org/smssecure/smssecure/ConversationActivity.java [hunk 2]
```diff
@@ -507,7 +513,7 @@ private void handleStartSecureSession() {
     builder.setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
       @Override
       public void onClick(DialogInterface dialog, int which) {
-        if (!isEncryptedConversation){
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/ConversationActivity.java]: same

File: src/org/smssecure/smssecure/ConversationActivity.java [hunk 3]
```diff
@@ -516,7 +522,7 @@ public void onClick(DialogInterface dialog, int which) {
         } else {
           allocatedThreadId = threadId;
         }
-        Log.w(TAG, "Refreshing thread "+allocatedThreadId+"...");
+        Log.w(TAG, "Refreshing thread " + allocatedThreadId + "...");
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/ConversationActivity.java]: same

File: src/org/smssecure/smssecure/ConversationActivity.java [hunk 4]
```diff
@@ -681,10 +687,20 @@ private void handleManualMmsRequired() {
...
     final Uri       draftMedia     = getIntent().getData();
+    final String    draftFilename  = getIntent().getStringExtra(FILENAME_EXTRA);
     final MediaType draftMediaType = MediaType.from(getIntent().getType());
 
     if (draftText != null)                            composeText.setText(draftText);
-    if (draftMedia != null && draftMediaType != null) setMedia(draftMedia, draftMediaType);
+    if (draftMedia != null && draftMediaType != null) {
+      Recipient    primaryRecipient       = getRecipients() == null ? null : getRecipients().getPrimaryRecipient();
+      boolean      isSecureSmsDestination = isSingleConversation() && SessionUtil.hasSession(this, masterSecret, primaryRecipient);
+      if(draftMediaType.equals(MediaType.FILE) && !isSecureSmsDestination){
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/ConversationActivity.java]: space between `if` and `(`

File: src/org/smssecure/smssecure/ConversationActivity.java [hunk 5]
```diff
@@ -681,10 +687,20 @@ private void handleManualMmsRequired() {
...
+        Toast.makeText(this, R.string.ConversationActivity_sorry_no_files_attaching_on_insecure_chat, Toast.LENGTH_LONG).show();
+      } else {
+        if(draftFilename != null) setMedia(draftMedia, draftMediaType, draftFilename);
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/ConversationActivity.java]: Can pull into the parent `if` as an `else if`

File: src/org/smssecure/smssecure/ConversationActivity.java [hunk 6]
```diff
@@ -723,7 +739,10 @@ protected void onPostExecute(List<Draft> drafts) {
             setMedia(Uri.parse(draft.getValue()), MediaType.AUDIO);
           } else if (draft.getType().equals(Draft.VIDEO)) {
             setMedia(Uri.parse(draft.getValue()), MediaType.VIDEO);
+          } else if (draft.getType().equals(Draft.FILE)) {
+            setMedia(Uri.parse(draft.getValue()) , MediaType.FILE);
           }
+
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/ConversationActivity.java]: remove whitespace

File: src/org/smssecure/smssecure/ConversationActivity.java [hunk 7]
```diff
@@ -966,16 +988,24 @@ public void onReceive(Context context, Intent intent) {
...
-      AttachmentManager.selectImage(this, PICK_IMAGE); break;
-    case AttachmentTypeSelectorAdapter.ADD_VIDEO:
-      AttachmentManager.selectVideo(this, PICK_VIDEO); break;
-    case AttachmentTypeSelectorAdapter.ADD_SOUND:
-      AttachmentManager.selectAudio(this, PICK_AUDIO); break;
-    case AttachmentTypeSelectorAdapter.ADD_CONTACT_INFO:
-      AttachmentManager.selectContactInfo(this, PICK_CONTACT_INFO); break;
-    case AttachmentTypeSelectorAdapter.TAKE_PHOTO:
-      attachmentManager.capturePhoto(this, TAKE_PHOTO); break;
+      case AttachmentTypeSelectorAdapter.ADD_IMAGE:
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/ConversationActivity.java]: Please keep previous formatting

File: src/org/smssecure/smssecure/ConversationItem.java [hunk 8]
```diff
@@ -525,7 +529,16 @@ public void onClick(final View v, final Slide slide) {
...
+      } else if (slide.hasFile()){
+        SaveAttachmentTask.showWarningDialog(context, new DialogInterface.OnClickListener() {
+          public void onClick(DialogInterface dialog, int which) {
+            SaveAttachmentTask saveTask = new SaveAttachmentTask(context, masterSecret);
+            saveTask.execute(new SaveAttachmentTask.Attachment(slide.getUri(), slide.getContentType(), new Date().getTime(), slide.getFileName()));
+            return;
+          }
+        });
+
+      }else{
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/ConversationItem.java]: `} else {`

File: src/org/smssecure/smssecure/ConversationItem.java [hunk 9]
```diff
@@ -525,7 +529,16 @@ public void onClick(final View v, final Slide slide) {
         intent.putExtra(MediaPreviewActivity.DATE_EXTRA, messageRecord.getTimestamp());
 
         context.startActivity(intent);
-      } else {
+      } else if (slide.hasFile()){
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/ConversationItem.java]: ...`hasfile()) {`

File: src/org/smssecure/smssecure/attachments/Attachment.java [hunk 10]
```diff
@@ -85,4 +87,7 @@ public void setThumbnail(@Nullable Bitmap thumbnail) {
   public Bitmap getThumbnail() {
     return thumbnail;
   }
+
+  @Nullable
+  public String getFileName(){ return fileName; }
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/attachments/Attachment.java]:
```java
public String getFileName() {
  return fileName; 
}
```

File: src/org/smssecure/smssecure/attachments/UriAttachment.java [hunk 11]
```diff
@@ -49,4 +56,28 @@ public boolean equals(Object other) {
...
+    }
+    if (fileName == null) {
+      fileName = uri.getPath();
+      int cut = fileName != null ? fileName.lastIndexOf('/') : -1;
+      if (cut != -1) {
+        fileName = fileName.substring(cut + 1);
+      }
+    }
+    return fileName;
+    }
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/attachments/UriAttachment.java]: unindent 1 level

File: src/org/smssecure/smssecure/database/AttachmentDatabase.java [hunk 12]
```diff
@@ -296,6 +298,7 @@ void insertAttachmentsForMessage(@NonNull MasterSecret masterSecret,
     DatabaseAttachment databaseAttachment = (DatabaseAttachment) attachment;
     File               dataFile           = getAttachmentDataFile(databaseAttachment.getAttachmentId(), DATA);
 
+
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/database/AttachmentDatabase.java]: remove whitespace

File: src/org/smssecure/smssecure/database/AttachmentDatabase.java [hunk 13]
```diff
@@ -305,6 +308,8 @@ void insertAttachmentsForMessage(@NonNull MasterSecret masterSecret,
     ContentValues contentValues = new ContentValues();
     contentValues.put(SIZE, dataSize);
     contentValues.put(CONTENT_TYPE, mediaStream.getMimeType());
+    if(attachment.getFileName() != null)
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/database/AttachmentDatabase.java]: Add space between `if` and `(``, use brackets

File: src/org/smssecure/smssecure/database/AttachmentDatabase.java [hunk 14]
```diff
@@ -463,6 +470,8 @@ private AttachmentId insertAttachment(MasterSecret masterSecret, long mmsId, Att
     contentValues.put(CONTENT_LOCATION, attachment.getLocation());
     contentValues.put(CONTENT_DISPOSITION, attachment.getKey());
     contentValues.put(NAME, attachment.getRelay());
+    if(attachment.getFileName() != null)
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/database/AttachmentDatabase.java]: Same as above

File: src/org/smssecure/smssecure/database/DatabaseFactory.java [hunk 15]
```diff
@@ -559,8 +564,8 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
 
       if (oldVersion < INTRODUCED_DRAFTS_VERSION) {
         db.execSQL("CREATE TABLE drafts (_id INTEGER PRIMARY KEY, thread_id INTEGER, type TEXT, value TEXT);");
-        executeStatements(db, new String[] {
-            "CREATE INDEX IF NOT EXISTS draft_thread_index ON drafts (thread_id);",
+        executeStatements(db, new String[]{
+                "CREATE INDEX IF NOT EXISTS draft_thread_index ON drafts (thread_id);",
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/database/DatabaseFactory.java]: unrelated to PR

File: src/org/smssecure/smssecure/database/DraftDatabase.java [hunk 16]
```diff
@@ -126,10 +127,11 @@ public String getValue() {
 
     public String getSnippet(Context context) {
       switch (type) {
-      case TEXT:  return value;
-      case IMAGE: return context.getString(R.string.DraftDatabase_Draft_image_snippet);
-      case VIDEO: return context.getString(R.string.DraftDatabase_Draft_video_snippet);
-      case AUDIO: return context.getString(R.string.DraftDatabase_Draft_audio_snippet);
+        case TEXT:  return value;
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/database/DraftDatabase.java]: leave previous formatting

File: src/org/smssecure/smssecure/database/MmsSmsDatabase.java [hunk 17]
```diff
@@ -172,12 +176,12 @@ private Cursor queryTables(String[] projection, String selection, String order,
 
     smsQueryBuilder.setTables(SmsDatabase.TABLE_NAME);
     mmsQueryBuilder.setTables(MmsDatabase.TABLE_NAME + " LEFT OUTER JOIN " +
-                              AttachmentDatabase.TABLE_NAME +
-                              " ON " + AttachmentDatabase.TABLE_NAME + "." + AttachmentDatabase.ROW_ID + " = " +
-                                  " (SELECT " + AttachmentDatabase.TABLE_NAME + "." + AttachmentDatabase.ROW_ID +
-                                  " FROM " + AttachmentDatabase.TABLE_NAME + " WHERE " +
-                                  AttachmentDatabase.TABLE_NAME + "." + AttachmentDatabase.MMS_ID + " = " +
-                                  MmsDatabase.TABLE_NAME + "." + MmsDatabase.ID + " LIMIT 1)");
+            AttachmentDatabase.TABLE_NAME +
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/database/MmsSmsDatabase.java]: Leave previous formatting

File: src/org/smssecure/smssecure/jobs/MmsSendJob.java [hunk 18]
```diff
@@ -234,6 +234,8 @@ private SendReq constructSendPdu(MasterSecret masterSecret, OutgoingMediaMessage
         PduPart part = new PduPart();
         part.setData(Util.readFully(PartAuthority.getAttachmentStream(context, masterSecret, attachment.getDataUri())));
         part.setContentType(Util.toIsoBytes(attachment.getContentType()));
+        if(attachment.getFileName() != null)
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/jobs/MmsSendJob.java]: space between `if` and `(`, add brackets

File: src/org/smssecure/smssecure/mms/AttachmentManager.java [hunk 19]
```diff
@@ -82,6 +80,24 @@ public AttachmentManager(@NonNull Activity activity, @NonNull AttachmentListener
     thumbnail.setOnClickListener(new ThumbnailClickListener());
   }
 
+  //public void clearIfFileSlides(){
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/AttachmentManager.java]: Remove

File: src/org/smssecure/smssecure/mms/AttachmentManager.java [hunk 20]
```diff
@@ -82,6 +80,24 @@ public AttachmentManager(@NonNull Activity activity, @NonNull AttachmentListener
...
+    //if(slideDeck.hasFileSlide()) {
+    //  slideDeck.clear();
+    //  attachmentView.setVisibility(View.GONE);
+    //  attachmentListener.onAttachmentChanged();
+    //}
+  //}
+
+
+  public void selectFile(final MasterSecret secret, final Activity activity, int requestCode) {
+        new FileChooser(activity).setFileListener(new FileChooser.FileSelectedListener() {
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/AttachmentManager.java]: correct indentation

File: src/org/smssecure/smssecure/mms/AttachmentManager.java [hunk 21]
```diff
@@ -92,7 +108,8 @@ public void onSuccess(Boolean result) {
       }
 
       @Override
-      public void onFailure(ExecutionException e) {}
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/AttachmentManager.java]: unrelated to PR

File: src/org/smssecure/smssecure/mms/AttachmentManager.java [hunk 22]
```diff
@@ -153,11 +178,11 @@ public void setMedia(@NonNull final MasterSecret masterSecret,
         long start = System.currentTimeMillis();
         try {
           final long  mediaSize = MediaUtil.getMediaSize(context, masterSecret, uri);
-          final Slide slide     = mediaType.createSlide(context, uri, mediaSize);
+          final Slide slide     = mediaType.createSlide(context, uri, mediaSize, fileName);
           Log.w(TAG, "slide with size " + mediaSize + " took " + (System.currentTimeMillis() - start) + "ms");
           return slide;
-        } catch (IOException ioe) {
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/AttachmentManager.java]: unrelated to PR

File: src/org/smssecure/smssecure/mms/AttachmentManager.java [hunk 23]
```diff
@@ -305,19 +330,20 @@ public void onClick(View v) {
...
+                                                String fileName)
+            throws IOException {
       switch (this) {
-      case IMAGE: return new ImageSlide(context, uri, dataSize);
-      case GIF:   return new GifSlide(context, uri, dataSize);
-      case AUDIO: return new AudioSlide(context, uri, dataSize);
-      case VIDEO: return new VideoSlide(context, uri, dataSize);
-      default:    throw  new AssertionError("unrecognized enum");
+        case IMAGE: return new ImageSlide(context, uri, dataSize);
+        case GIF:   return new GifSlide(context, uri, dataSize);
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/AttachmentManager.java]: keep previous indentation

File: src/org/smssecure/smssecure/mms/AttachmentManager.java [hunk 24]
```diff
@@ -305,19 +330,20 @@ public void onClick(View v) {
...
+    IMAGE, GIF, AUDIO, VIDEO, FILE;
 
     public @NonNull Slide createSlide(@NonNull Context context,
                                       @NonNull Uri     uri,
-                                               long    dataSize)
-        throws IOException
-    {
+                                               long    dataSize,
+                                                String fileName)
+            throws IOException {
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/AttachmentManager.java]: keep previous indentation

File: src/org/smssecure/smssecure/mms/AttachmentTypeSelectorAdapter.java [hunk 25]
```diff
@@ -71,22 +75,35 @@ public View getView(int position, View convertView, ViewGroup parent) {
...
       addItem(data, context.getString(R.string.AttachmentTypeSelectorAdapter_camera),  ResUtil.getDrawableRes(context, R.attr.conversation_attach_camera), TAKE_PHOTO);
     }
     addItem(data, context.getString(R.string.AttachmentTypeSelectorAdapter_picture), ResUtil.getDrawableRes(context, R.attr.conversation_attach_image),        ADD_IMAGE);
     addItem(data, context.getString(R.string.AttachmentTypeSelectorAdapter_video),   ResUtil.getDrawableRes(context, R.attr.conversation_attach_video),        ADD_VIDEO);
-    addItem(data, context.getString(R.string.AttachmentTypeSelectorAdapter_audio),   ResUtil.getDrawableRes(context, R.attr.conversation_attach_sound),        ADD_SOUND);
+    addItem(data, context.getString(R.string.AttachmentTypeSelectorAdapter_audio), ResUtil.getDrawableRes(context, R.attr.conversation_attach_sound), ADD_SOUND);
     addItem(data, context.getString(R.string.AttachmentTypeSelectorAdapter_contact), ResUtil.getDrawableRes(context, R.attr.conversation_attach_contact_info), ADD_CONTACT_INFO);
-
+      if(isEncryptedConversation && SMSSecurePreferences.getMultipartMMS(context)) {
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/AttachmentTypeSelectorAdapter.java]: space between `if` and `(`, fix indentation

File: src/org/smssecure/smssecure/mms/AttachmentTypeSelectorAdapter.java [hunk 26]
```diff
@@ -71,22 +75,35 @@ public View getView(int position, View convertView, ViewGroup parent) {
...
+                      fileIdentifier = new IconListItem(context.getString(R.string.AttachmentTypeSelectorAdapter_file), ResUtil.getDrawableRes(context, R.attr.conversation_attach), ADD_FILE);
+                data.add(fileIdentifier);
+              }
     return data;
   }
 
   private static void addItem(List<IconListItem> list, String text, int resource, int id) {
     list.add(new IconListItem(text, resource, id));
   }
+    public void setSecureDestination(boolean isSecure){
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/AttachmentTypeSelectorAdapter.java]: Fix indentation

File: src/org/smssecure/smssecure/mms/FileChooser.java [hunk 27]
```diff
@@ -0,0 +1,134 @@
...
+import android.view.WindowManager.LayoutParams;
+import android.widget.AdapterView;
+import android.widget.ArrayAdapter;
+import android.widget.ListView;
+import android.widget.TextView;
+import java.io.File;
+import java.io.FileFilter;
+import java.util.Arrays;
+
+public class FileChooser {
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/FileChooser.java]: Make indentation in this file conform to the rest of the project

File: src/org/smssecure/smssecure/mms/FileChooser.java [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+    private final Activity activity;
+    private ListView list;
+    private Dialog dialog;
+    private File currentPath;
+
+    // filter on file extension
+    private String extension = null;
+    public void setExtension(String extension) {
+        this.extension = (extension == null) ? null :
+                extension.toLowerCase();
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/FileChooser.java]: can go on same line

File: src/org/smssecure/smssecure/mms/FileSlide.java [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+import org.smssecure.smssecure.util.ResUtil;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+
+import ws.com.google.android.mms.ContentType;
+import ws.com.google.android.mms.pdu.PduPart;
+
+public class FileSlide extends Slide {
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/FileSlide.java]: Fix indentation for file

File: src/org/smssecure/smssecure/mms/MmsMediaConstraints.java [hunk 30]
```diff
@@ -38,4 +38,7 @@ public int getVideoMaxSize() {
   public int getAudioMaxSize() {
     return MAX_MESSAGE_SIZE;
   }
+
+  @Override
+  public int getFileMaxSize() { return MAX_MESSAGE_SIZE; }
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/MmsMediaConstraints.java]: put function contents on a new line

File: src/org/smssecure/smssecure/mms/OutgoingMediaMessage.java [hunk 31]
```diff
@@ -1,7 +1,9 @@
 package org.smssecure.smssecure.mms;
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/OutgoingMediaMessage.java]: Seems like nothing in this file needs to change?

File: src/org/smssecure/smssecure/mms/Slide.java [hunk 32]
```diff
@@ -38,7 +38,6 @@
   public Slide(@NonNull Context context, @NonNull Attachment attachment) {
     this.context    = context;
     this.attachment = attachment;
-
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/Slide.java]: unrelated to PR

File: src/org/smssecure/smssecure/mms/Slide.java [hunk 33]
```diff
@@ -72,6 +71,10 @@ public boolean hasAudio() {
     return false;
   }
 
+  public boolean hasFile() { return false; }
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/Slide.java]: function contents on new lines

File: src/org/smssecure/smssecure/mms/Slide.java [hunk 34]
```diff
@@ -102,10 +105,21 @@ public boolean hasPlaceholder() {
...
+  }
+  protected static Attachment constructAttachmentFromUri(@NonNull Context context,
+                                                         @NonNull Uri     uri,
+                                                         @NonNull String  defaultMime,
+                                                         long     size,
+                                                         String   fileName)
   {
     Optional<String> resolvedType = Optional.fromNullable(MediaUtil.getMimeType(context, uri));
-    return new UriAttachment(uri, resolvedType.or(defaultMime), AttachmentDatabase.TRANSFER_PROGRESS_STARTED, size);
+    if(fileName != null)
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/Slide.java]: space between `if` and `(` and add brackets

File: src/org/smssecure/smssecure/mms/Slide.java [hunk 35]
```diff
@@ -118,6 +132,7 @@ public boolean equals(Object other) {
            this.hasAudio() == that.hasAudio()                        &&
            this.hasImage() == that.hasImage()                        &&
            this.hasVideo() == that.hasVideo()                        &&
+           this.hasFile() == that.hasFile()                          &&
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/Slide.java]: line up with line above

File: src/org/smssecure/smssecure/mms/SlideDeck.java [hunk 36]
```diff
@@ -86,7 +86,29 @@ public void addSlide(Slide slide) {
...
+      if (slide.hasImage() || slide.hasVideo() || slide.hasAudio() || slide.hasFile()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public boolean removeFileSlides() {
+    boolean returnObject = false;
+    if(slides.isEmpty())
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/mms/SlideDeck.java]: space and brackets

File: src/org/smssecure/smssecure/util/MediaUtil.java [hunk 37]
```diff
@@ -88,6 +90,8 @@ public static Slide getSlideForAttachment(Context context, Attachment attachment
              ? ContentType.IMAGE_JPEG
              : mimeType;
     default:
+      if(ContentType.isNonTextVideoImageAudioType(mimeType))
```
Comment (pR0Ps) [inline src/org/smssecure/smssecure/util/MediaUtil.java]: space and brackets

**Comment by pR0Ps** [comment]
@vdveer Thanks for this!

I didn't really have too much time to look at the actual content at the moment, but I highlighted a few coding style issues.

In general, leave things the way they are if possible and follow the coding style of the surrounding code (brackets, indentation, etc)

**Comment by vdveer** [comment]
@pR0Ps Do you use some kind of formatting plugin or is there any way to include the formatting profile in the repository? A lot of the formatting issues where made by some form of autoformatting.

**Comment by pR0Ps** [comment]
I typically just use Android Studi, IIRC there is an option in it to match
the existing formatting. I also make sure to just go over the diff manually
and look for formatting issues before committing it.

@BastienLQ: Do you have a different strategy?

**Comment by BastienLQ** [comment]
I use Atom and it formats code automagically based on previous indentation.

**Comment by BastienLQ** [comment]
@vdveer Can you rebase your commit on `upstream`?

**Comment by vdveer** [comment]
@BastienLQ Sure, I will get to that a little later. In the meantime a question; when reading contribution policy for the whispersystems upstream, I read the words 'think twice when adding a user preference'. In this PR I introduced a preference to activate filetransfer. I introduced this so the user is sure what he wants when he applies the option, also because when a user sends a file to a user with an older version of the app will receive a message without file attachment (text-only when text is also added). What do you think is the best way to go here? Leave the preference or remove it and give everyone the file option in the menu on secure threads?

**Comment by BastienLQ** [comment]
Actually, as I just merged `upstream` into `master`, can you rebase your branch on `SMSSecure:master`?
You can remove the preference you added.

**Comment by vdveer** [comment]
@BastienLQ removed preference, rebased on master, squashed and forced as new head on this PR's branch.

**Comment by BastienLQ** [comment]
When a text file is sent, message is displayed as a file for the sender, but as a text MMS (with encoding issues) for the recipient.

I made some fixes in 633b910 / [`file-transfert`](https://github.com/SMSSecure/SMSSecure/commits/file-transfert) (indentation and assets). Can you base your modifications on this commit? I can also add you in the SMSSecure organization if you want.

**Comment by vdveer** [comment]
That would safe me soms rebasing, so that would Be great. I will look into the text file issues tonight or the weekend. I had a merge conflict there with upstream changes, so I'll do some more testing.

**Pull request closed without merging**